salt/tests: Add unit testing for Salt Jinja templates

gdemonet opened the pull request:
**Component**: salt, tests

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: See #3073

**Summary**:

This PR introduces unit tests for all Jinja templates under `salt/`.
The goal is to automatically ensure that rendering logic is valid and will not break at runtime.

At this stage, no coverage information is provided. Furthermore, only rendering is verified, and the output is simply discarded - a future PR may add optional validity checks for the output (e.g. Salt states as YAML).

**Acceptance criteria**:

- All templates are tested as part of `tox -e unit-tests`, and in CI


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3073

---

<details>
  <summary>Task list</summary>

- [x] Document unit testing strategy
- [x] Implement basic logic for automatically testing all templates
- [x] Implement logic for configuring rendering contexts
- [x] Progressively enable tests for all templates:
    - [x] metalk8s.addons
    - [x] metalk8s.archives
    - [x] metalk8s.beacon
    - [x] metalk8s.container-engine
    - [x] metalk8s.internal
    - [x] metalk8s.kubectl
    - [x] metalk8s.kubernetes
    - [x] metalk8s.node
    - [x] metalk8s.orchestrate
    - [x] metalk8s.reactor
    - [x] metalk8s.repo
    - [x] metalk8s.roles
    - [x] metalk8s.salt
    - [x] metalk8s.service-configuration
    - [x] metalk8s.solutions
    - [x] metalk8s.sreport
    - [x] metalk8s.utils
    - [x] metalk8s.volumes
</details>

gdemonet committed:
docs: Document Salt formulas unit testing

We also re-organize the "Development" chapter within the "Developer
Guide" book, to separate "Testing" from "Best Practices".

See: #3073

gdemonet committed:
build/versions: Extract Salt versions.json data

This data is moved from the target used to build the
`salt/metalk8s/versions.json` file to the `buildchain.versions` module,
for easier access in tests.

See: #3073

gdemonet committed:
salt: Rename Jinja templates with .j2 extension

This will ease the discovery of templates for testing.

See: #3073

gdemonet committed:
build(style): Rework formatting for Black

The `buildchain.salt_tree` was formatted with Black in #3071, but some
strings were still (wrongly) split.
This makes things look nicer.

See: #2650
See: #3071

gdemonet committed:
salt/tests: Setup 'unit.formulas' test skeleton

Include pylint, mypy, and yamllint hooks configuration for these
modules, before we start implementing any logic.
The test functionality is absent from this commit, and will be added in
later.

See: #3073

gdemonet committed:
salt/tests: Setup minimal Jinja environment fixture

This basic environment is configured to behave as Salt does at runtime.
It will read templates from the sources, under the `salt/` directory.

See: #3073

gdemonet committed:
salt/tests: List all Jinja templates

We expose, and use, a `template_path` fixture, which is parameterized to
generate one test case per Jinja template in the salt/ directory.

Running `tox -e unit-tests -- -m formulas` shows:
- 93 passed
- 143 xfailed

Interestingly, this means more than a third of the `.sls`/`.sls.in`/...
files are "simple" (or empty) Jinja templates, which don't require any
available context to render (e.g. many `init.sls` are only containing an
`include` list, without rendering logic).

See: #3073

gdemonet committed:
salt/tests: Enable rendering context configuration

A configuration file, at `salt/tests/unit/formulas/config.yaml`,
describes which templates to validate, and how.
This commit introduces the logic for parameterizing the rendering
context through configuration options, and demonstrates its most minimal
usage by skipping all formulas under "metalk8s" directory.

Running `tox -e unit-tests -- -m formulas` shows:
- 235 skipped
- 1 xfailed (the only formula which doesn't live under metalk8s/, which
is top.sls.in)

See: #3073

gdemonet committed:
salt/tests: Add base `grains` to rendering context

We extract (and clean up) grains from a MetalK8s 2.7 bootstrap node,
running on CentOS 7.9.
These are stored in a YAML data source, which is read and exposed to the
rendering tests in a "base context" fixture.

Now, the only template which was not skipped, `top.sls.in`, can be
rendered without errors.
We can now expect that all non-skipped templates can be rendered with
their configured options, and update the test accordingly.

Running `tox -e unit-tests -- -m formulas` shows:
- 1 passed
- 235 skipped

See: #3073

gdemonet committed:
salt/tests: Add tests for rendering map.jinja

This commit adds all required mocks to be able to test the map.jinja
template, which is imported from many other templates.

We first define a common mock of the `salt` object to make Salt
execution modules available in the rendering context. This mock stores
references of some parts of the context (grains, pillar, environment) to
define "smarter" mocks of execution methods, such as the
`grains.filter_by` method which needs to rely on active grains,
available as `salt_mock._grains`.

We also enrich the base context with an example pillar, retrieved from a
MetalK8s 2.7.1 bootstrap node.

Finally, to make the rendering test more exhaustive, we define an "os"
configuration option, which allows to update grains to simulate running
on different OS distributions/releases. All templates will be rendered on
"CentOS/7" by default, but map.jinja will be rendered with all supported
OS values.

Running `tox -e unit-tests -- -m formulas` shows:
- 2 passed
- 234 skipped

See: #3073

gdemonet committed:
salt/tests: Render reactor SLS with extra context

We define a new "extra_context" configuration option, which allows to
add arbitrary data to the rendering context.
This is especially useful for some specific formulas, like Reactors,
which expect a `data` value to contain the event for which they are
called.

Running `tox -e unit-tests -- -m formulas` shows:
- 3 passed
- 233 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering of metalk8s.archives.*

This required adding a mock of `metalk8s.get_archives`.

Running `tox -e unit-tests -- -m formulas` shows:
- 6 passed
- 230 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering of metalk8s.beacon.*

Nothing to add for these formulas (it's mostly using map.jinja).

Running `tox -e unit-tests -- -m formulas` shows:
- 7 passed
- 229 skipped

See: #3073

gdemonet committed:
salt/tests: Check metalk8s.container-engine.*

For these formulas, we need to add the `slspath` variable in context
(which depends on the template we are rendering), and define some extra
context for the containerd/files/50-metalk8s.conf.j2 file.

Running `tox -e unit-tests -- -m formulas` shows:
- 14 passed
- 222 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering of metalk8s.internal.*

Added two static mocks for `mount.swaps` and `network.connect`.

Running `tox -e unit-tests -- -m formulas` shows:
- 20 passed
- 216 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.kubectl.*

Nothing to add in fixtures.

Running `tox -e unit-tests -- -m formulas` shows:
- 23 passed
- 213 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.kubernetes.*

Add a "saltenv" configuration option, with a smart default to
"metalk8s-{{ pillar.metalk8s.cluster_version }}".
Add another "pillar_overrides" config option, to pass in some specific
pillar values to update without replacing all the existing pillar
contents.

Since we now have more options, we rework the representation of a
context on test failures.

Add a minimal mock for Kubernetes API, currently only enabling "get" requests against
v1/Node.

Add mocks for:
- hashutil.base64_b64decode
- metalk8s.format_san
- metalk8s.minions_by_role
- metalk8s_kubernetes.get_object
- metalk8s_network.get_cluster_dns_ip
- metalk8s_network.get_kubernetes_service_ip
- metalk8s_network.get_mtu_from_ip
- mine.get
- pillar.get

Running `tox -e unit-tests -- -m formulas` shows:
- 84 passed
- 152 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.node.*

Add a "minion_state" configuration option, to allow simulation of
various minion states, such as "ready", the default described by
"data/base_*.yaml" files, or "new", used to simulate a fresh minion
installation, when bootstrapping or expanding.

Since this new option mutates a big chunk of the `grains` dict, we
realized test contexts were not independent. This commit changes this
to ensure no leakage between tests is possible.

Add a mock for metalk8s_network.get_ip_from_cidrs.

Running `tox -e unit-tests -- -m formulas` shows:
- 85 passed
- 151 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.repo.*

Add a mock for file.join, and some static context expected by some file
templates.

Running `tox -e unit-tests -- -m formulas` shows:
- 96 passed
- 140 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.roles.*

Nothing to add.

Running `tox -e unit-tests -- -m formulas` shows:
- 112 passed
- 124 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.salt.*

Add mocks for config.get and metalk8s_solutions.list_available, with some extra context
variables to render file templates.

Also add a "standalone" minion state, where some pillar keys are not available and
some changes to the default minion configuration.

Running `tox -e unit-tests -- -m formulas` shows:
- 129 passed
- 107 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.service-configuration.*

Running `tox -e unit-tests -- -m formulas` shows:
- 130 passed
- 106 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.solutions.*

`metalk8s.solutions.available` is a good example of a formula which
needs such unit tests, and we use "pillar_overrides" to render it with all
expected branches. If these need to be shared with other templates, they may
become captured in a dedicated configuration option.

Add a mock for metalk8s_solutions.manifest_from_iso to infer reasonable
manifests from archive paths.

Running `tox -e unit-tests -- -m formulas` shows:
- 132 passed
- 104 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.sreport.*

Running `tox -e unit-tests -- -m formulas` shows:
- 134 passed
- 102 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.utils.*

Running `tox -e unit-tests -- -m formulas` shows:
- 137 passed
- 99 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.volumes.*

Add a new "volumes" configuration option to control the
metalk8s:volumes" pillar using simple aliases.

Running `tox -e unit-tests -- -m formulas` shows:
- 141 passed
- 95 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.dex.*

We add a "k8s_overrides" configuration option to flexibly adjust the
default objects in our K8s API mock (see data/kubernetes/base.yaml).
This allows us to test all branches for
metalk8s.addons.dex.deployed.service-configuration.

Add mocks for:
- file.read
- hashutil.base64_encodefile
- metalk8s_network.get_oidc_service_ip
- metalk8s_service_configuration.get_service_conf
- slsutil.renderer

Running `tox -e unit-tests -- -m formulas` shows:
- 155 passed
- 81 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.logging.*

Running `tox -e unit-tests -- -m formulas` shows:
- 169 passed
- 67 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.nginx-ingress.*

Running `tox -e unit-tests -- -m formulas` shows:
- 178 passed
- 58 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.nginx-ingress-control-plane.*

Add a "mode" option to control whether the formulas are rendered on a
minion or a master. For this option to have any effect, we add an `opts`
object to context, and customize it with this option value.

This allows us to cover all branches for
metalk8s.addons.nginx-ingress-control-plane.control-plane-ip.

Add a mock for saltutil.runner (which naively forwards calls to the
relevant mocked functions), and update the mine.get mock to allow
retrieving mine data for remote minions (no assumption on `tgt`).

Running `tox -e unit-tests -- -m formulas` shows:
- 184 passed
- 52 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.prometheus-adapter.*

Running `tox -e unit-tests -- -m formulas` shows:
- 186 passed
- 50 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.prometheus-operator.*

Add support for listing objects in K8s API mock, and add a mock for
metalk8s_grafana.load_dashboard.

Running `tox -e unit-tests -- -m formulas` shows:
- 197 passed
- 39 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.addons.solutions.*

Running `tox -e unit-tests -- -m formulas` shows:
- 200 passed
- 36 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for remaining metalk8s.addons.*

Running `tox -e unit-tests -- -m formulas` shows:
- 210 passed
- 26 skipped

See: #3073

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

gdemonet committed:
lint: Bump mypy to v0.812

Update buildchain code to be compatible.

gdemonet committed:
lint: Fix regex for non-Salt pylint targets

Buildchain code was not checked.

gdemonet committed:
salt/tests: Check rendering for metalk8s.orchestrate.bootstrap.*

We expose the `opts` context variable in `SaltMock` instances, to
access it from mocks, and enforce that `saltutil` mocks are called
within a Salt master context.
We also add a map of minions to `SaltMock`, to help implement a mock
for `saltutil.cmd`, and enable future enhancements (potentially using
new/reworked configuration options).

Also add a mock for `grains.get`.

Running `tox -e unit-tests -- -m formulas` shows:
- 214 passed
- 22 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.orchestrate.certs.*

Define some test cases depending on custom `pillar.orchestrate` values.
A useful next step could be to extract and verify the values of some
computed Jinja variables (`sls` and `post_mods`). A project for another
day.

Running `tox -e unit-tests -- -m formulas` shows:
- 215 passed
- 21 skipped

See: #3073

alexandre-allard commented on salt/tests/unit/formulas/config.yaml:
```diff
@@ -0,0 +1,582 @@
+
+metalk8s:
+  # Use the special `_skip` keyword to omit rendering of a directory or formula
+  _skip: true
```
I would rather have `_skip` set to `false` at the top level (or even as a default) and explicitly skip the things we don't want to/can't test, but maybe it's already what is planned when everything will be ready.

gdemonet replied:
`_skip` is already set to `false` by default, but since I wanted to build this PR by progressively adding more formulas to the scope, one of the first commits "skips everything except `map.jinja`" and the following ones are iteratively "unskipping" parts of the `metalk8s/` tree.

The goal is to get rid of the `_skip` statements after this PR, so nothing is skipped.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/3073-add-jinja-unit-tests` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `feature/3073-add-jinja-unit-tests`
* Rebase `feature/3073-add-jinja-unit-tests` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

gdemonet committed:
salt/tests: Rework configuration of test cases

The previous approach was based on generating a Cartesian product of all
passed options, but this can create issues when some values of different
options are inherently incompatible (e.g. passing `pillar_overrides` for
adjusting knowledge about a node which doesn't exist in some
`architecture`).

We thus rework the approach to declare test cases directly, optionally
providing the ability to combine "option sets" through hierarchy.
For more details, the developer documentation has been updated to better
illustrate the new principles for tests configuration.

See: #3073

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

gdemonet committed:
salt/tests: Check rendering for metalk8s.orchestrate.downgrade.*

Add an "architecture" configuration option to simply define a set of
minions, following one of the supported deployment architectures.
This option will handle updating the mocked minions, the mocked K8s
API, and the mocked pillar data, accordingly.

Adjust the "saltenv" configuration option to be more deterministic.
The previous approach allowed a special "__default__" value, which was
replaced by whatever value was available in
`pillar.metalk8s.cluster_version`. The new approach instead sets this
pillar value (and the version of all nodes in `pillar.metalk8s.nodes`)
from the version passed as a saltenv, and expects other options to
override these versions when desired.

Add mock for `pkg.version_cmp` and `metalk8s.cmp_sorted`.

Running `tox -e unit-tests -- -m formulas` shows:
- 219 passed
- 17 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.orchestrate.solutions.*

Add mock for `file.find` (which never finds anything, for now, as it
has no impact on the rendering logic).

Running `tox -e unit-tests -- -m formulas` shows:
- 226 passed
- 10 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for metalk8s.orchestrate.upgrade.*

Same principles as for metalk8s.orchestrate.downgrade.* formulas.

Running `tox -e unit-tests -- -m formulas` shows:
- 230 passed
- 6 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for rest of metalk8s.orchestrate.*

Introduce `BaseOption.PRIORITY` to help managing execution order for
options when building a rendering context.
Update the `MinionState` "new" value to adapt which minion is referred to
depending on the `MinionMode` value (master or minion).

Running `tox -e unit-tests -- -m formulas` shows:
- 234 passed
- 2 skipped

See: #3073

gdemonet committed:
salt/tests: Check rendering for rest of metalk8s.*

Remove all mentions of `_skip` (except a comment for readers) from the
configuration file.

Running `tox -e unit-tests -- -m formulas` shows:
- 236 passed
- 0 skipped

Closes: #3073

alexandre-allard reviewed (approved):
Really cool stuff you did there.
I tried some common mistakes in Jinja and it works well.
LGTM.

gdemonet commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



Please check the status of the associated issue None.

Goodbye gdemonet.
